storage/client/client.go

```diff
@@ -0,0 +1,1081 @@
+type StorageClient struct {
+	db     storage.TargetStorage
+	logger *log.Logger
+}
+
+// NewStorageClient creates a new storage client.
+func NewStorageClient(db storage.TargetStorage, l *log.Logger) *StorageClient {
+	return &StorageClient{db, l}
+}
+
+// Shutdown closes the backing TargetStorage.
+func (c *StorageClient) Shutdown() {
+	c.db.Shutdown()
+}
+
+// Status returns status information for the Oasis Indexer.
+func (c *StorageClient) Status(ctx context.Context) (*Status, error) {
+	qf := NewQueryFactory(strcase.ToSnake(LatestChainID))
+
+	s := Status{
+		LatestChainID: LatestChainID,
+	}
+	if err := c.db.QueryRow(
+		ctx,
+		qf.StatusQuery(),
+	).Scan(&s.LatestBlock, &s.LatestUpdate); err != nil {
+		c.logger.Info("row scan failed",
+			"request_id", ctx.Value(RequestIDContextKey),
+			"err", err.Error(),
+		)
+		return nil, common.ErrStorageError
+	}
+	return &s, nil
+}
+
+// Blocks returns a list of consensus blocks.
+func (c *StorageClient) Blocks(ctx context.Context, r *BlocksRequest, p *common.Pagination) (*BlockList, error) {
+	cid, ok := ctx.Value(ChainIDContextKey).(string)
+	if !ok {
+		return nil, common.ErrBadChainID
+	}
+	qf := NewQueryFactory(cid)
+
+	rows, err := c.db.Query(
+		ctx,
+		qf.BlocksQuery(),
+		r.From,
+		r.To,
+		r.After,
+		r.Before,
+		p.Limit,
+		p.Offset,
+	)
+	if err != nil {
+		c.logger.Info("query failed",
+			"request_id", ctx.Value(RequestIDContextKey),
+			"err", err.Error(),
+		)
+		return nil, common.ErrStorageError
+	}
+	defer rows.Close()
+
+	bs := BlockList{
+		Blocks: []Block{},
+	}
+	for rows.Next() {
+		var b Block
+		if err := rows.Scan(&b.Height, &b.Hash, &b.Timestamp); err != nil {
+			c.logger.Info("row scan failed",
+				"request_id", ctx.Value(RequestIDContextKey),
+				"err", err.Error(),
+			)
+			return nil, common.ErrStorageError
+		}
+		b.Timestamp = b.Timestamp.UTC()
+
+		bs.Blocks = append(bs.Blocks, b)
+	}
+
+	return &bs, nil
+}
+
+// Block returns a consensus block.
+func (c *StorageClient) Block(ctx context.Context, r *BlockRequest) (*Block, error) {
+	cid, ok := ctx.Value(ChainIDContextKey).(string)
+	if !ok {
+		return nil, common.ErrBadChainID
+	}
+	qf := NewQueryFactory(cid)
+
+	var b Block
+	if err := c.db.QueryRow(
+		ctx,
+		qf.BlockQuery(),
+		r.Height,
+	).Scan(&b.Height, &b.Hash, &b.Timestamp); err != nil {
+		c.logger.Info("row scan failed",
+			"request_id", ctx.Value(RequestIDContextKey),
+			"err", err.Error(),
+		)
+		return nil, common.ErrStorageError
+	}
+	b.Timestamp = b.Timestamp.UTC()
+
+	return &b, nil
+}
+
+// Transactions returns a list of consensus transactions.
+func (c *StorageClient) Transactions(ctx context.Context, r *TransactionsRequest, p *common.Pagination) (*TransactionList, error) {
+	cid, ok := ctx.Value(ChainIDContextKey).(string)
+	if !ok {
+		return nil, common.ErrBadChainID
+	}
+	qf := NewQueryFactory(cid)
+
+	rows, err := c.db.Query(
+		ctx,
+		qf.TransactionsQuery(),
+		r.Block,
+		r.Method,
+		r.Sender.String(),
```

Comment:
Not sure whether all these args are automatically string-ified. I _think_ they are but I made it explicit just in case.

Reply:
Confirmed that they are automatically serialized properly. Removed the `String()` because these fields are optional and can trigger nil pointer exception errors if left in